pkg/tools/clientmap/refcount.go

```diff
@@ -31,54 +31,75 @@ type refcountClient struct {
 }
 
 func (r *refcountClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
-	return r.client.Request(ctx, request)
+	return r.client.Request(ctx, request, opts...)
 }
 
 func (r *refcountClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
-	return r.client.Close(ctx, conn)
+	return r.client.Close(ctx, conn, opts...)
 }
 
 // RefcountMap - clientmap.Map wrapped with refcounting.  Each Store,Load,LoadOrStore increments the count,
 //               each Delete decrements the count.  When the count is zero, Delete deletes.
 type RefcountMap struct {
-	Map
+	m Map
 }
 
 // Store sets the value for a key.
 // Increments the refcount
 func (r *RefcountMap) Store(key string, value networkservice.NetworkServiceClient) {
 	client := &refcountClient{client: value}
-	r.Map.Store(key, client)
+	r.m.Store(key, client)
 	atomic.AddInt32(&client.count, 1)
 }
 
-// LoadOrStore returns the existing value for the key if present. Otherwise, it stores and returns the given value. The loaded result is true if the value was loaded, false if stored.
+// LoadOrStore returns the existing value for the key if present. Otherwise, it stores and returns the given value. The
+// loaded result is true if the value was loaded, false if stored.
 // Increments the refcount.
 func (r *RefcountMap) LoadOrStore(key string, value networkservice.NetworkServiceClient) (networkservice.NetworkServiceClient, bool) {
-	client := &refcountClient{client: value}
-	rv, loaded := r.Map.LoadOrStore(key, client)
-	if client, ok := rv.(*refcountClient); ok {
-		atomic.AddInt32(&client.count, 1)
-		return client.client, loaded
-	}
-	return rv, loaded
+	raw, loaded := r.m.LoadOrStore(key, &refcountClient{client: value})
+	client := raw.(*refcountClient)
+
+	atomic.AddInt32(&client.count, 1)
+
+	return client.client, loaded
 }
 
-// Load returns the value stored in the map for a key, or nil if no value is present. The ok result indicates whether value was found in the map.
+// Load returns the value stored in the map for a key, or nil if no value is present. The ok result indicates whether
+// value was found in the map.
 // Increments refcount.
 func (r *RefcountMap) Load(key string) (networkservice.NetworkServiceClient, bool) {
-	rv, loaded := r.Map.Load(key)
-	if client, ok := rv.(*refcountClient); ok {
-		atomic.AddInt32(&client.count, 1)
-		return client.client, loaded
+	raw, loaded := r.m.Load(key)
+	if !loaded {
+		return nil, false
+	}
+	client := raw.(*refcountClient)
+
+	atomic.AddInt32(&client.count, 1)
+
+	return client.client, loaded
+}
+
+// LoadAndDelete returns the value stored in the map for a key, and decrements the refcount. If no value found or the
+// refcount < 0, returns (nil, false). If refcount == 0, deletes the value.
+func (r *RefcountMap) LoadAndDelete(key string) (networkservice.NetworkServiceClient, bool) {
```

Comment:
Note: LoadAndDelete must be done under one _lock section_. Please try to use `r.m.LoadAndDelete`

Reply:
Yep, there was an issue. `r.m.LoadAndDelete` can't be used here because refcount is stored in the map entry. Added `sync.RWMutex` to fix issue.